game-backend/src/MainController.ts

```diff
@@ -48,26 +51,54 @@ export default class MainController {
     matchUpdate: MatchUpdate
   ) => void {
     return (userId: UserId, matchUpdate: MatchUpdate): void => {
-      this.store.userContexts.userContextMap[userId].forEach(ws => {
+      this.store.userConnnections.userConnectionMap[userId].forEach(ws => {
         ws.send(JSON.stringify(matchUpdate));
       });
     };
   }
 
   private getTerminateExecutor(): () => void {
     return (): void => {
+      // ping all websocket connections
       setTimeout(() => {
+        this.removeInactiveUsers();
+
         this.gameController = new GameController(
           this.getMatchUpdateExecutor(),
           this.getTerminateExecutor()
         );
-        this.store.userContexts.userContextIds.forEach(id => {
+        this.store.userConnnections.userConnectionIds.forEach(id => {
           this.gameController.appendUser(id);
         });
       }, GAME_RESTART_TIME);
     };
   }
 
+  /**
+   * Remove users from store if there is not at least one connection alive
+   * @private
+   * @memberof MainController
+   */
+  private removeInactiveUsers(): void {
+    this.store.userConnnections.userConnectionIds.forEach(userId => {
+      const foundAliveConnections = this.store.userConnnections.userConnectionMap[
+        userId
+      ].filter((ws: WebSocketWithStatus) => {
+        return ws.isAlive;
```

Comment:
Checking for `isAlive` here will result in unfair timing behavior.
Given: healtheck reinitiated a check at `xx:xx:00`, this means that all connections will be marked as `!isAlive`. When the game restarts at e.g. `xx:xx:01` (which is less time than the healthcheck would permit to reply to a ping) it will remove connections even though they will probably still be alive.
To fix this checking for the `readyState` might be more appropriate: https://github.com/websockets/ws/blob/master/doc/ws.md#ready-state-constants

Reply:
Yeah, that's a good solution.